### PR TITLE
server: fix span use after Finish() 

### PR DESCRIPTION
**Pull request by andreimatei**
The closure serving a pgwire connection was capturing a context with a
long-lived span, and so all connections were logging to that span.
That was bad, but what was even worse is that the long-lived span can
end before the closures, on server shutdown. That was
use-after-Finish(), which is currently reluctantly tolerated, but won't
be tolerated much longer.

This patch fixes it by giving each connection its own span.

Touches #58164

Release note: None

### CODE REVIEW

File: pkg/util/netutil/net.go [hunk 1]
```diff
@@ -150,12 +153,15 @@ func (s *Server) ServeWith(
 			return e
 		}
 		tempDelay = 0
-		go func() {
+		err := stopper.RunAsyncTask(ctx, "pgwire-serve", func(ctx context.Context) {
 			defer stopper.Recover(ctx)
 			s.Server.ConnState(rw, http.StateNew) // before Serve can return
-			serveConn(rw)
+			serveConn(ctx, rw)
 			s.Server.ConnState(rw, http.StateClosed)
```
Comment (knz) [inline pkg/util/netutil/net.go:160]: might be worth `defer` on the 2nd ConnState call here, since youre in this area.

**Review by knz — APPROVED** [review]

**Comment by knz** [comment]
cc @dhartunian for learning purposes (maybe you can ventilate this in your team for future reference)

**Review by andreimatei — COMMENTED** [review]
TFTR

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72492#-:-Mo-9XKF5ZVIUnB5cYyb:b-4as1hm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @knz)

___
*[pkg/util/netutil/net.go, line 160 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/72492#-Mnm2QumoGIi62PVA8Oo-r4-160:-Mo-9TlD9d5rA98zEZaH:b-1sg17r) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d9404f881c69c7f1ff8a8514904ec62fa169024/pkg/util/netutil/net.go#L160)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

might be worth `defer` on the 2nd ConnState call here, since youre in this area.
</blockquote></details>

will do it separately

**Review by andreimatei — COMMENTED** [review]
bors r-

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72492#-:-Mo-PiFl4CH1QvuJ_JWT:b-ydi9aw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @knz)

**Review by andreimatei — COMMENTED** [review]
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72492#-:-Mo-rFjG1AQX-y-eTJkj:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @knz)

**Review by andreimatei — COMMENTED** [review]
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72492#-:-Mo0uNax6W5fl76qJ2Se:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @knz)

**Pull request merged**